<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 2: Predicting Housing Prices

--- 
# Part 4: Clean test.csv

--- 

Don't want to drop ANY rows of test.csv only columns
* Fill NaNs with whatever same methods I used on the training data
* Add any engineered features and dummified columns if applicable!

**Rows should stay the same!**

**Columns should match final cleaned trained data**

### Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import missingno as msno
from scipy.stats import norm, skew

### Importing the data for this project

In [2]:
# Cleaned up the test csv in a new notebook by calling it the same df as before
ames = pd.read_csv('./datasets/test.csv')

### Display the first 5 rows of the dataframe

In [3]:
ames.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0  2658  902301120          190        RM          69.0      9142   Pave   
1  2718  905108090           90        RL           NaN      9662   Pave   
2  2414  528218130           60        RL          58.0     17104   Pave   
3  1989  902207150           30        RM          60.0      8520   Pave   
4   625  535105100           20        RL           NaN      9500   Pave   

  Alley Lot Shape Land Contour  ... 3Ssn Porch Screen Porch Pool Area Pool QC  \
0  Grvl       Reg          Lvl  ...          0            0         0     NaN   
1   NaN       IR1          Lvl  ...          0            0         0     NaN   
2   NaN       IR1          Lvl  ...          0            0         0     NaN   
3   NaN       Reg          Lvl  ...          0            0         0     NaN   
4   NaN       IR1          Lvl  ...          0          185         0     NaN   

  Fence Misc Feature Misc Val Mo Sold  Yr Sold  Sale Type  
0   NaN          NaN        0       4     2006        WD   
1   NaN          NaN        0       8     2006        WD   
2   NaN          NaN        0       9     2006        New  
3   NaN          NaN        0       7     2007        WD   
4   NaN          NaN        0       7     2009        WD   

[5 rows x 80 columns]

### Checking # of Rows and Columns

In [4]:
ames.shape

(878, 80)

In [5]:
#creating a df with just the categorical columns
categorical_df = ames.select_dtypes(include=['object'])

In [6]:
#creating a df with just the numeric columns
numeric_df =ames.select_dtypes(exclude=['object'])

### Fixing some data types.

#### Numeric to Categorical

In [7]:
# #MS SubClass (building class)
# all_data['MS SubClass'] = all_data['MS SubClass'].apply(str)

# #Overall Cond (1-10 but it's really categorical)
# all_data['Overall Cond'] = all_data['Overall Cond'].astype(str)

# #Overall Qual (1-10 but it's really categorical)
# all_data['Overall Qual'] = all_data['Overall Qual'].astype(str)

# #Year and month sold should be str and then dummified
# all_data['Yr Sold'] = all_data['Yr Sold'].astype(str)
# all_data['Mo Sold'] = all_data['Mo Sold'].astype(str)

#### Objects to Numeric

In [8]:
# cols = ('FireplaceQu', 'BsmtQual', 'BsmtCond', 'GarageQual', 'GarageCond', 
#         'ExterQual', 'ExterCond','HeatingQC', 'PoolQC', 'KitchenQual', 'BsmtFinType1', 
#         'BsmtFinType2', 'Functional', 'Fence', 'BsmtExposure', 'GarageFinish', 'LandSlope',
#         'LotShape', 'PavedDrive', 'Street', 'Alley', 'CentralAir', 'MSSubClass', 'OverallCond', 
#         'YrSold', 'MoSold')

### Drop Unnecessary Rows and Columns and Fill Missing Data If Applicable

In [9]:
#Drop PID columns from each
ames.drop('PID', axis = 1, inplace = True)

In [10]:
# Fill the NaNs in 'Alley','Fireplace Qu','Pool QC','Fence' columns with None
fill_columns = ['Alley','Fireplace Qu','Pool QC','Fence','Misc Feature']
ames[fill_columns]= ames[fill_columns].fillna('None')
ames.head(2)

Id  MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley Lot Shape  \
0  2658          190        RM          69.0      9142   Pave  Grvl       Reg   
1  2718           90        RL           NaN      9662   Pave  None       IR1   

  Land Contour Utilities  ... 3Ssn Porch Screen Porch Pool Area Pool QC Fence  \
0          Lvl    AllPub  ...          0            0         0    None  None   
1          Lvl    AllPub  ...          0            0         0    None  None   

  Misc Feature Misc Val  Mo Sold  Yr Sold  Sale Type  
0         None        0        4     2006        WD   
1         None        0        8     2006        WD   

[2 rows x 79 columns]

In [11]:
#Drop column(s) missing majority of data where None isn't meaningful to use
#Drop Misc Feature and therefore drop Misc Val since it pertains to other Misc column
# ames = ames.drop(['Misc Feature', 'Misc Val'],axis=1)
# ames.head(2)

In [12]:
ames.isnull().sum().sort_values().tail(22)

Exter Cond          0
Exter Qual          0
Year Built          0
Exterior 2nd        0
Exterior 1st        0
Sale Type           0
Roof Style          0
Roof Matl           0
Mas Vnr Area        1
Electrical          1
Mas Vnr Type        1
BsmtFin Type 2     25
Bsmt Qual          25
Bsmt Cond          25
Bsmt Exposure      25
BsmtFin Type 1     25
Garage Type        44
Garage Qual        45
Garage Cond        45
Garage Finish      45
Garage Yr Blt      45
Lot Frontage      160
dtype: int64

In [13]:
# Fill the NaNs in categorical Bsmt columns with 'None' as NA is missing when you look at value counts
fill_columns = ['Bsmt Qual','Bsmt Cond','Bsmt Exposure','BsmtFin Type 1','BsmtFin Type 2']
ames[fill_columns]= ames[fill_columns].fillna('None')

In [14]:
# Fill the NaNs in categorical Garage columns with 'None' as NA is missing when you look at value counts
fill_columns = ['Garage Type','Garage Finish','Garage Qual','Garage Cond']
ames[fill_columns]= ames[fill_columns].fillna('None')
ames.isnull().sum().sort_values().tail(13)

Exter Cond         0
Foundation         0
Exterior 2nd       0
Exterior 1st       0
Roof Matl          0
Roof Style         0
Exter Qual         0
Sale Type          0
Mas Vnr Area       1
Mas Vnr Type       1
Electrical         1
Garage Yr Blt     45
Lot Frontage     160
dtype: int64

In [15]:
#Drop Mas Vnr Type and Mas Vnr Area columns - These are Masonry veneer type & area which are not usually important to sale
#ames = ames.drop(['Mas Vnr Type', 'Mas Vnr Area'],axis=1)


#Maybe will drop these two columns eventually but for now we can just fill with None for NaN
fill_columns = ['Mas Vnr Type', 'Mas Vnr Area']
ames[fill_columns]= ames[fill_columns].fillna('None')
ames.head(2)

Id  MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley Lot Shape  \
0  2658          190        RM          69.0      9142   Pave  Grvl       Reg   
1  2718           90        RL           NaN      9662   Pave  None       IR1   

  Land Contour Utilities  ... 3Ssn Porch Screen Porch Pool Area Pool QC Fence  \
0          Lvl    AllPub  ...          0            0         0    None  None   
1          Lvl    AllPub  ...          0            0         0    None  None   

  Misc Feature Misc Val  Mo Sold  Yr Sold  Sale Type  
0         None        0        4     2006        WD   
1         None        0        8     2006        WD   

[2 rows x 79 columns]

In [16]:
# Replacing missing data for Garage Yr Blt, Garage Area and Garage Cars with 0 since no garage means 0 area and cars
fill_columns = ['Garage Yr Blt','Garage Area','Garage Cars']
ames[fill_columns] = ames[fill_columns].fillna(0)

In [17]:
#Fill NaNs for Lot Frontage with the median Lot Frontage of that neighborhood.
ames["Lot Frontage"] = ames.groupby("Neighborhood")["Lot Frontage"].transform(lambda x: x.fillna(x.median()))

In [18]:
ames.isnull().sum().sort_values().tail(12)

Total Bsmt SF    0
Foundation       0
Exter Qual       0
Mas Vnr Area     0
Mas Vnr Type     0
Exterior 2nd     0
Exterior 1st     0
Roof Matl        0
Roof Style       0
Exter Cond       0
Sale Type        0
Electrical       1
dtype: int64

In [19]:
# Fill rest of the Lot Frontage and Bsmt columns with 0 since they're numeric and probably don't have a basement 
# and might not have lot frontage in those neighborhoods
fill_columns = ['Lot Frontage','BsmtFin SF 2', 'BsmtFin SF 1', 'Bsmt Unf SF', 'Total Bsmt SF', 'Bsmt Half Bath', 'Bsmt Full Bath']
ames[fill_columns] = ames[fill_columns].fillna(0)
ames.isnull().sum().sort_values().tail(2)

Sale Type     0
Electrical    1
dtype: int64

In [20]:
# Checking I still have 878 rows
ames.shape

(878, 79)

### Drop Columns that have Low Variation in Box Plots
Street, Utilities, Condition 2, Roof Matl, Heating, Electrical

Exterior 1st, Bsmt Qual, Garage Qual??

In [21]:
ames.shape

(878, 79)

In [22]:
ames = ames.drop(['Electrical','Heating','Roof Matl','Condition 2','Utilities','Street'],axis=1)
ames.shape

(878, 73)

## Feature Engineering

In [23]:
#Create new column of the House Age
ames['House Age']= (ames['Yr Sold']- ames['Year Built'])
ames['House Age'].describe()

count    878.000000
mean      37.333713
std       30.469253
min        0.000000
25%        8.250000
50%       36.000000
75%       54.750000
max      129.000000
Name: House Age, dtype: float64

In [24]:
# Create Total SF (total house square footage) column as this is usually a key metric for house sales
ames['Total SF'] = ames['Total Bsmt SF'] + ames['1st Flr SF'] + ames['2nd Flr SF']
ames['Total SF'].describe()

count     878.000000
mean     2528.543280
std       772.173406
min       720.000000
25%      1968.000000
50%      2431.500000
75%      2974.750000
max      6872.000000
Name: Total SF, dtype: float64

In [25]:
# Create Total Bath column as this is usually a key metric for house sales
ames['Total Bath'] = ames['Full Bath'] + ames['Half Bath'] + ames['Bsmt Full Bath']+ ames['Bsmt Half Bath']
ames['Total Bath'].describe()

count    878.000000
mean       2.433941
std        0.936811
min        1.000000
25%        2.000000
50%        2.000000
75%        3.000000
max        8.000000
Name: Total Bath, dtype: float64

In [26]:
#Create Total Porch SF
ames['Porch'] = ames['Open Porch SF'] + ames['Enclosed Porch']+ ames['3Ssn Porch']+ ames['Screen Porch']
ames['Porch'].describe()

count     878.000000
mean       88.563781
std       112.496111
min         0.000000
25%         0.000000
50%        50.000000
75%       131.750000
max      1207.000000
Name: Porch, dtype: float64

In [27]:
# Drop any columns used to make new features as this can cause data leakage
ames = ames.drop(['Full Bath', 'Half Bath', 'Bsmt Full Bath', 'Bsmt Half Bath', 'Total Bsmt SF', '1st Flr SF',
                   '2nd Flr SF','Yr Sold','Year Built','Open Porch SF','Enclosed Porch','3Ssn Porch','Screen Porch'],axis=1)

### Drop columns that had categories that didn't match the test.csv data
When we dummified these it did not create them in the test data when we were running through the same cleaning/feature engineering steps

In [28]:
ames.columns

Index(['Id', 'MS SubClass', 'MS Zoning', 'Lot Frontage', 'Lot Area', 'Alley',
       'Lot Shape', 'Land Contour', 'Lot Config', 'Land Slope', 'Neighborhood',
       'Condition 1', 'Bldg Type', 'House Style', 'Overall Qual',
       'Overall Cond', 'Year Remod/Add', 'Roof Style', 'Exterior 1st',
       'Exterior 2nd', 'Mas Vnr Type', 'Mas Vnr Area', 'Exter Qual',
       'Exter Cond', 'Foundation', 'Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure',
       'BsmtFin Type 1', 'BsmtFin SF 1', 'BsmtFin Type 2', 'BsmtFin SF 2',
       'Bsmt Unf SF', 'Heating QC', 'Central Air', 'Low Qual Fin SF',
       'Gr Liv Area', 'Bedroom AbvGr', 'Kitchen AbvGr', 'Kitchen Qual',
       'TotRms AbvGrd', 'Functional', 'Fireplaces', 'Fireplace Qu',
       'Garage Type', 'Garage Yr Blt', 'Garage Finish', 'Garage Cars',
       'Garage Area', 'Garage Qual', 'Garage Cond', 'Paved Drive',
       'Wood Deck SF', 'Pool Area', 'Pool QC', 'Fence', 'Misc Feature',
       'Misc Val', 'Mo Sold', 'Sale Type', 'House Age', 'Total 

In [29]:
ames = ames.drop(['Bsmt Cond', 'Exterior 1st', 'Exterior 2nd', 'Functional', 'Garage Qual', 'Heating QC', 'Mas Vnr Area', 
                  'Misc Feature', 'MS Zoning', 'Pool QC', 'Sale Type', 'Mas Vnr Type'],axis=1)
ames.shape

(878, 52)

In [30]:
# Make sure 2 neighborhoods are not in the test dataset: 'GrnHill' & 'Landmrk'
ames['Neighborhood'].unique()

array(['OldTown', 'Sawyer', 'Gilbert', 'NAmes', 'MeadowV', 'CollgCr',
       'Somerst', 'Mitchel', 'SawyerW', 'IDOTRR', 'BrkSide', 'Edwards',
       'ClearCr', 'NWAmes', 'Timber', 'NoRidge', 'NridgHt', 'Crawfor',
       'StoneBr', 'Veenker', 'BrDale', 'SWISU', 'Blmngtn', 'NPkVill',
       'Greens', 'Blueste'], dtype=object)

In [31]:
# Either replace 'VWD' as this is not in the train dataset or drop the whole column
#ames['Sale Type'].value_counts()

In [32]:
ames = ames.drop(ames[(ames['Kitchen Qual'] == 'Po')].index)

In [33]:
ames.shape

(877, 52)

In [34]:
# Commenting out as I did not drop these for the Production Model
# Another round of dropping columns after running the Lasso Model

# ames = ames.drop(['Bsmt Unf SF','Alley', 'Lot Shape', 'Land Contour', 'Lot Config', 'Land Slope','Condition 1',
#                  'House Style','Exter Qual','Foundation', 'Bsmt Qual', 'BsmtFin Type 1', 'BsmtFin Type 2', 
#                   'Garage Type','Garage Finish'],axis=1)
# ames.shape

### Dummifying Categorical Features


In [35]:
#recalling the categorical df to make sure the edits we made to the ames df carry over
categorical_df =ames.select_dtypes(include=['object'])
categorical_df.columns

Index(['Alley', 'Lot Shape', 'Land Contour', 'Lot Config', 'Land Slope',
       'Neighborhood', 'Condition 1', 'Bldg Type', 'House Style', 'Roof Style',
       'Exter Qual', 'Exter Cond', 'Foundation', 'Bsmt Qual', 'Bsmt Exposure',
       'BsmtFin Type 1', 'BsmtFin Type 2', 'Central Air', 'Kitchen Qual',
       'Fireplace Qu', 'Garage Type', 'Garage Finish', 'Garage Cond',
       'Paved Drive', 'Fence'],
      dtype='object')

In [36]:
ames.shape

(877, 52)

In [37]:
 ames = pd.get_dummies(ames, columns=categorical_df.columns)

In [38]:
ames.shape

(877, 177)

878 Rows expected by Kaggle

177 Columns is what we should have - 1 less than the training dataset with 178

## Final Cleaning

### Clean up column names.

In [39]:
# Convert all columns to lowercase and replace spaces in column names.
ames.columns = ames.columns.str.lower().str.replace(' ', '_')
ames.columns

Index(['id', 'ms_subclass', 'lot_frontage', 'lot_area', 'overall_qual',
       'overall_cond', 'year_remod/add', 'bsmtfin_sf_1', 'bsmtfin_sf_2',
       'bsmt_unf_sf',
       ...
       'garage_cond_po', 'garage_cond_ta', 'paved_drive_n', 'paved_drive_p',
       'paved_drive_y', 'fence_gdprv', 'fence_gdwo', 'fence_mnprv',
       'fence_mnww', 'fence_none'],
      dtype='object', length=177)

### Save your cleaned dataframe as csv file


In [40]:
#export cleaned dataset to save
ames.to_csv('./datasets/Ames_Test_Clean.csv', index = False)